### PR TITLE
data state restoration

### LINKED ISSUE #63: Data state restoration
Currently, the `--train_from` option does not include means of restoring corpora states, hence training resumes from the beginning of the bitexts. This entails resumed models are training on a subset of the available data, unless some manual shuffling is done between each resumption.

Fixes (partially implemented on V2, but never ported here
- [x] add a line index as files are read, pass it along when collating batches, and skip up til this line index upon training resumption
- [ ] a more complex refactoring is needed to save the dataloader state, as it would involve communicating all examples in the reservoir &mdash; which would be much more costly to communicate.

CC @jrvc

### PR DESCRIPTION
**Pull request by TimotheeMickus**
closes #63 .

same idea as v2, didn't bother porting from it.

does not embark bucket states, although this could maybe be done by picking the line indices from all examples in the reservoir. Would come at a (potentially high) communication overhead cost.

### CODE REVIEW

**Comment by TimotheeMickus** [comment]
also embarks a minor debug on update_vocab being only partially removed (following #61)

File: mammoth/inputters/dataset.py [hunk 1]
```diff
@@ -162,13 +184,13 @@ def _cast(example_dict):
                 if self.transforms is not None else lambda x: x
             ),
             stride=self.stride,
-            offset=self.offset,
+            offset=offset,
         )
         examples = map(_cast, examples)
         yield from examples
 
     # FIXME: some RNN archs require sorting src's by length
```
Comment (jrvc) [inline mammoth/inputters/dataset.py:192]: Out of this PR concenrs... do we still support RNNs? if not, we can just delete this

File: mammoth/models/model_saver.py [hunk 2]
```diff
@@ -95,12 +96,14 @@ def save(self, step, moving_average=None):
                 self._rm_checkpoint(todel)
             self.checkpoint_queue.append(chkpt_names)
 
-    def _save(self, step):
+    def _save(self, step, save_model, data_state, device_context):
         """Save a resumable checkpoint.
 
         Args:
             step (int): step number
             model (nn.Module): torch model to save
```
Comment (jrvc) [inline mammoth/models/model_saver.py:104]: missing: save_model (type): description

**Review by jrvc — APPROVED** [review]
lgtm :)

File: mammoth/inputters/dataset.py (hunk 1 above)
Comment (TimotheeMickus) [reply mammoth/inputters/dataset.py:192]: no, we no longer support RNNs. the move is to externalize transformer variantsvia lucidrains (see #56). Will remove the comment

File: mammoth/models/model_saver.py (hunk 2 above)
Comment (TimotheeMickus) [reply mammoth/models/model_saver.py:104]: yeah, i think it was renamed from model to save_model? not sure why. will fix that.
